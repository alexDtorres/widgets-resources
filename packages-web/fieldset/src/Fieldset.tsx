import { createElement, ReactNode } from "react";
import { hot } from "react-hot-loader/root";

import { FieldsetContainerProps } from "../typings/FieldsetProps";

const Fieldset = (props: FieldsetContainerProps): ReactNode => {
    const { legend, content, name, tabIndex, style, class: className } = props;

    return (
        <fieldset name={name} tabIndex={tabIndex} style={style} className={className}>
            {legend?.value && <legend>{legend.value}</legend>}
            {content}
        </fieldset>
    );
};

export default hot(Fieldset);
