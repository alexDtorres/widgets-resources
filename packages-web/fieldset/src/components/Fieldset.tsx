import { createElement, CSSProperties, FunctionComponent } from "react";

export interface FieldsetProps {
    name?: string;
    className: string;
    style?: CSSProperties;
    tabIndex?: number;
    legend?: string;
}

export const Fieldset: FunctionComponent<FieldsetProps> = (props): JSX.Element | null => {
    const { legend, name, tabIndex, style, className, children } = props;

    if (!children) {
        return null;
    }

    return (
        <fieldset name={name} tabIndex={tabIndex} style={style} className={className}>
            {legend && <legend>{legend}</legend>}
            {children}
        </fieldset>
    );
};
