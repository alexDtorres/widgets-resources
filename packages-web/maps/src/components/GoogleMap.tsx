import { createElement, CSSProperties, ReactElement, useEffect, useRef, useState } from "react";
import classNames from "classnames";

import googleApiWrapper from "./GoogleApi";
import { Alert } from "@widgets-resources/piw-utils";
import { getDimensions } from "../utils";
import { Marker, SharedProps } from "../../typings";

export interface GoogleMapsProps extends SharedProps {
    scriptsLoaded?: boolean;
    validationMessage?: string;
    mapsToken?: string;
    className?: string;
    mapStyles?: string;
    divStyles?: CSSProperties;

    optionStreetView: boolean;
    mapTypeControl: boolean;
    fullScreenControl: boolean;
    rotateControl: boolean;
}

export const GoogleMap = (props: GoogleMapsProps): ReactElement => {
    const map = useRef<google.maps.Map>();
    const googleMapsRef = useRef<HTMLDivElement>(null);
    const defaultCenterLocation: google.maps.LatLngLiteral = { lat: 51.9066346, lng: 4.4861703 };
    let bounds!: google.maps.LatLngBounds;

    const [markers, setMarkers] = useState<google.maps.Marker[]>([]); //Used to manage and remove markers from the map
    const [validationMessage, setValidationMessage] = useState(props.validationMessage);

    useEffect(() => {
        if (props.scriptsLoaded) {
            createUpdateMap();
        }
    }, []);

    useEffect(() => {
        createUpdateMap();
    }, [props.locations, props.currentLocation]);

    const createUpdateMap = (): void => {
        const mapOptions: google.maps.MapOptions = {
            zoom: props.zoomLevel,
            zoomControl: props.optionZoomControl,
            scrollwheel: props.optionScroll,
            draggable: props.optionDrag,
            streetViewControl: props.optionStreetView,
            mapTypeControl: props.mapTypeControl,
            fullscreenControl: props.fullScreenControl,
            rotateControl: props.rotateControl,
            minZoom: 1,
            maxZoom: 20,
            styles: getMapStyles()
        };
        if (googleMapsRef.current && !map.current) {
            map.current = new google.maps.Map(googleMapsRef.current, mapOptions);
        } else if (map.current) {
            map.current.setOptions(mapOptions);
        }
        addMarkers();
    };

    const addMarkers = (): void => {
        if (props.locations && props.locations.length > 0) {
            markers.forEach(marker => marker.setMap(null));
            setMarkers([]);
            bounds = new google.maps.LatLngBounds();
            setMarkers(
                props.locations.reduce<google.maps.Marker[]>((markerArray, currentLocation) => {
                    const marker = addMarker(currentLocation);
                    if (marker) {
                        markerArray.push(marker);
                    }
                    return markerArray;
                }, [])
            );
        }
        addCurrentLocation();
        updateCamera();
    };

    const addCurrentLocation = () => {
        if (props.currentLocation) {
            const currentLocationMarker = addMarker(props.currentLocation);
            if (currentLocationMarker) {
                markers.push(currentLocationMarker);
            }
        }
    };

    const addMarker = (marker: Marker): google.maps.Marker | undefined => {
        if (map.current) {
            bounds.extend({
                lat: marker.latitude,
                lng: marker.longitude
            });
            const mapMarker = new google.maps.Marker({
                position: {
                    lat: marker.latitude,
                    lng: marker.longitude
                },
                icon: marker.url,
                title: marker.title
            });
            if (marker.title) {
                const infoContent = document.createElement("span");
                infoContent.innerHTML = marker.title || "";
                if (marker.onClick) {
                    infoContent.style.cursor = "pointer";
                    infoContent.onclick = marker.onClick;
                }
                const infoWindow = new google.maps.InfoWindow({
                    content: infoContent
                });
                mapMarker.addListener("click", () => {
                    infoWindow.open(map.current, mapMarker);
                });
            } else {
                if (marker.onClick) {
                    mapMarker.addListener("click", () => {
                        marker.onClick!();
                    });
                } else {
                    mapMarker.setClickable(false);
                }
            }
            mapMarker.setMap(map.current);
            return mapMarker;
        }
        return undefined;
    };

    const updateCamera = (): void => {
        const { zoomLevel, autoZoom } = props;
        setTimeout(() => {
            if (bounds && map.current) {
                try {
                    if (!autoZoom) {
                        map.current.setCenter({
                            lat: markers[0].getPosition()?.lat() ?? defaultCenterLocation.lat,
                            lng: markers[0].getPosition()?.lng() ?? defaultCenterLocation.lng
                        });
                        map.current.setZoom(zoomLevel);
                    } else {
                        map.current.fitBounds(bounds);
                    }
                } catch (error) {
                    setValidationMessage(`Invalid map bounds ${error.message}`);
                }
            }
        }, 0);
    };

    const getMapStyles = (): google.maps.MapTypeStyle[] => {
        if (props.mapStyles && props.mapStyles.trim()) {
            try {
                return JSON.parse(props.mapStyles);
            } catch (error) {
                setValidationMessage(`invalid Map styles, ${error.message}`);
            }
        }

        return [
            {
                featureType: "poi",
                elementType: "labels",
                stylers: [{ visibility: "off" }]
            }
        ];
    };

    return (
        <div
            className={classNames("widget-maps", props.className)}
            style={{ ...props.divStyles, ...getDimensions(props) }}
        >
            {validationMessage && (
                <Alert bootstrapStyle="danger" className="widget-google-maps-alert">
                    {validationMessage}
                </Alert>
            )}
            <div className="widget-google-maps-wrapper">
                <div className="widget-google-maps" ref={googleMapsRef} />
            </div>
        </div>
    );
};

export default googleApiWrapper("https://maps.googleapis.com/maps/api/js?key=")(GoogleMap);
