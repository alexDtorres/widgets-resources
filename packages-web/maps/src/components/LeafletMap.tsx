import { createElement, ReactElement, useEffect, useRef, useState } from "react";
import {
    FeatureGroup,
    LatLngLiteral,
    Map,
    MapOptions,
    Marker as LeafletMarker,
    icon,
    tileLayer,
    TileLayerOptions,
    CRS
} from "leaflet";
import { MapProviderEnum, Marker, SharedProps } from "../../typings";
import { Alert } from "@widgets-resources/piw-utils";
import classNames from "classnames";
import { customUrls, getDimensions, mapAttr } from "../utils";

export interface LeafletProps extends SharedProps {
    mapProvider: MapProviderEnum;
    attributionControl: boolean;
}

export const LeafletMap = (props: LeafletProps): ReactElement => {
    const map = useRef<Map>();
    const leafletRef = useRef<HTMLDivElement>(null);
    const defaultCenterLocation: LatLngLiteral = { lat: 51.9066346, lng: 4.4861703 };
    const markerGroup = new FeatureGroup();

    const [validationMessage, setValidationMessage] = useState(props.validationMessage);

    useEffect(() => {
        createUpdateMap();
    }, []);

    useEffect(() => {
        createUpdateMap();
    }, [props.locations, props.currentLocation]);

    const createUpdateMap = (): void => {
        const mapOptions: MapOptions = {
            scrollWheelZoom: props.optionScroll,
            zoomControl: props.optionZoomControl,
            attributionControl: props.attributionControl,
            zoom: props.zoomLevel,
            minZoom: 1,
            maxZoom: 20,
            // Work around for page scroll down to bottom on first click on map in Chrome and IE
            // https://github.com/Leaflet/Leaflet/issues/5392
            keyboard: false,
            dragging: props.optionDrag,
            crs: CRS.EPSG4326
        };
        if (leafletRef.current && !map.current) {
            map.current = new Map(leafletRef.current, mapOptions).addLayer(baseMapLayer());
        } else if (map.current) {
            //TODO: Test it
            map.current.options = mapOptions;
            map.current.setZoom(props.zoomLevel);
        }
        addMarkers();
    };

    const addMarkers = (): void => {
        if (map.current && markerGroup) {
            //TODO: Implementation of marker logic
            markerGroup.clearLayers();
            if (props.locations && props.locations.length) {
                props.locations.forEach(location => {
                    const marker = addMarker(location);
                    const layer = markerGroup.addLayer(marker);
                    map.current!.addLayer(layer);
                });
            }
            addCurrentLocation();
            updateCamera();
        }
    };

    const addMarker = (marker: Marker): LeafletMarker => {
        const mapMarker = new LeafletMarker([Number(marker.latitude), Number(marker.longitude)]);
        if (marker.title) {
            mapMarker.setTooltipContent(marker.title);
            if (marker.onClick) {
                const infoContent = document.createElement("span");
                infoContent.innerHTML = marker.title || "";
                if (marker.onClick) {
                    infoContent.style.cursor = "pointer";
                    infoContent.onclick = marker.onClick;
                }
                mapMarker.bindPopup(infoContent).openPopup();
            } else {
                mapMarker.bindPopup(marker.title).openPopup();
            }
        } else {
            if (marker.onClick) {
                mapMarker.on("click", () => {
                    marker.onClick!();
                });
            }
        }
        if (marker.url) {
            mapMarker.setIcon(
                icon({
                    iconUrl: marker.url,
                    iconSize: [32, 32],
                    className: "marker"
                })
            );
        }
        return mapMarker;
    };

    const updateCamera = (): void => {
        const { zoomLevel, autoZoom } = props;
        setTimeout(() => {
            if (map.current && markerGroup && markerGroup.getLayers().length > 0) {
                const bounds = markerGroup.getBounds();
                try {
                    if (!autoZoom) {
                        map.current.panTo(getCenter(), { animate: false });
                        map.current.setZoom(zoomLevel);
                    } else {
                        if (bounds.isValid()) {
                            map.current.fitBounds(bounds, { animate: false }).invalidateSize();
                        }
                    }
                } catch (error) {
                    setValidationMessage(`Invalid map bounds ${error.message}`);
                }
            }
        }, 0);
    };

    const getCenter = (): LatLngLiteral => {
        return {
            lat: props.locations?.[0]?.latitude ?? defaultCenterLocation.lat,
            lng: props.locations?.[0]?.longitude ?? defaultCenterLocation.lng
        };
    };

    const addCurrentLocation = () => {
        if (map.current && props.currentLocation) {
            const currentLocationMarker = addMarker(props.currentLocation);
            if (currentLocationMarker) {
                const layer = markerGroup.addLayer(currentLocationMarker);
                map.current.addLayer(layer);
            }
        }
    };

    const baseMapLayer = () => {
        const { mapProvider, mapsToken } = props;

        let urlTemplate = "";
        let mapAttribution = "";
        if (mapProvider === "mapBox") {
            urlTemplate = customUrls.mapbox + mapsToken;
            mapAttribution = mapAttr.mapboxAttr;
        } else if (mapProvider === "hereMaps" && mapsToken && mapsToken.indexOf(",") > 0) {
            const splitToken = mapsToken.split(",");
            urlTemplate = customUrls.hereMaps + `app_id=${splitToken[0]}&app_code=${splitToken[1]}`;
            mapAttribution = mapAttr.hereMapsAttr;
        } else {
            urlTemplate = customUrls.openStreetMap;
            mapAttribution = mapAttr.openStreetMapAttr;
        }

        return tileLayer(urlTemplate, {
            attribution: mapAttribution,
            id: mapProvider === "mapBox" ? "mapbox.streets" : undefined
        } as TileLayerOptions);
    };

    return (
        <div className={classNames("widget-maps", props.className)} style={{ ...props.style, ...getDimensions(props) }}>
            {validationMessage && (
                <Alert bootstrapStyle="danger" className="widget-leaflet-maps-alert leaflet-control">
                    {validationMessage}
                </Alert>
            )}
            <div className="widget-leaflet-maps-wrapper">
                <div className="widget-leaflet-maps" ref={leafletRef} />
            </div>
        </div>
    );
};
