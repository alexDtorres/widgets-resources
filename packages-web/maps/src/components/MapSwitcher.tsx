import { createElement, ReactElement } from "react";
import "leaflet/dist/leaflet.css";
import "leaflet-defaulticon-compatibility";
import "leaflet-defaulticon-compatibility/dist/leaflet-defaulticon-compatibility.webpack.css";
import "../ui/Maps.css";
import { GoogleMap, GoogleMapsProps } from "./GoogleMap";
import { LeafletMap, LeafletProps } from "./LeafletMap";

interface SwitcherProps extends GoogleMapsProps, LeafletProps {}

export const MapSwitcher = (props: SwitcherProps): ReactElement => {
    if (props.mapProvider === "googleMaps") {
        return <GoogleMap {...props} />;
    }
    return <LeafletMap {...props} />;
};
