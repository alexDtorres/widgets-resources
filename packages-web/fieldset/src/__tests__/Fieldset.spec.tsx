import { shallow } from "enzyme";
import { shallowToJson } from "enzyme-to-json";
import { createElement, Fragment, ReactNode } from "react";

import { Fieldset, FieldsetProps } from "../components/Fieldset";

// export interface FieldsetProps {
//     name?: string;
//     className: string;
//     style?: CSSProperties;
//     tabIndex?: number;
//     legend?: string;
// }

describe("Fieldset", () => {
    const defaultFieldsetProps: FieldsetProps = {
        name: "fieldset",
        className: "className",
        style: {},
        tabIndex: 0,
        legend: "legend"
    };

    const defaultChildren: ReactNode = (
        <Fragment>
            <label>Name:</label>
            <br />
            <input type="text" name="employee_name" id="employee_name" />
        </Fragment>
    );

    it("renders children and legend", () => {
        const fieldset = shallow(<Fieldset {...defaultFieldsetProps}>{defaultChildren}</Fieldset>);

        expect(shallowToJson(fieldset)).toMatchSnapshot();
    });
    it("renders only children when no legend is passed", () => {
        const fieldset = shallow(
            <Fieldset {...defaultFieldsetProps} legend={undefined}>
                {defaultChildren}
            </Fieldset>
        );

        expect(shallowToJson(fieldset)).toMatchSnapshot();
    });
    it("renders nothing when no children are passed", () => {
        const fieldset = shallow(<Fieldset {...defaultFieldsetProps} />);

        expect(shallowToJson(fieldset)).toMatchSnapshot();
    });
});
