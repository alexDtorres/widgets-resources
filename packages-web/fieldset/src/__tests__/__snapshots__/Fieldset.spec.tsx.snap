// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fieldset renders children and legend 1`] = `
<fieldset
  className="className"
  name="fieldset"
  style={Object {}}
  tabIndex={0}
>
  <legend>
    legend
  </legend>
  <label>
    Name:
  </label>
  <br />
  <input
    id="employee_name"
    name="employee_name"
    type="text"
  />
</fieldset>
`;

exports[`Fieldset renders nothing when no children are passed 1`] = `""`;

exports[`Fieldset renders only children when no legend is passed 1`] = `
<fieldset
  className="className"
  name="fieldset"
  style={Object {}}
  tabIndex={0}
>
  <label>
    Name:
  </label>
  <br />
  <input
    id="employee_name"
    name="employee_name"
    type="text"
  />
</fieldset>
`;
